### PR TITLE
S3-Mock: add local heap storage

### PR DESCRIPTION
**Pull request by snazy**
This allows us to omit starting a test container for simple S3 test cases.

### CODE REVIEW

File: testing/s3mock/src/main/java/org/projectnessie/s3mock/S3Resource.java [hunk 1]
```diff
@@ -362,6 +370,66 @@ public Response getObject(
...
+            return Response.status(405, "PUT object not allowed").build();
+          }
+        });
+  }
+
+  private InputStream chunkedInput(String contentEncoding, InputStream input) throws IOException {
+    if (contentEncoding != null) {
+      List<String> contentEncodingList =
+          Arrays.stream(contentEncoding.split(",")).collect(Collectors.toList());
+      if (contentEncodingList.remove("identity")) {
```
Comment (adutra) [inline testing/s3mock/src/main/java/org/projectnessie/s3mock/S3Resource.java:423]: `Collectors.toList()` offers no mutability guarantee. Calling `remove` on the returned list seems risky.

File: testing/s3mock/src/main/java/org/projectnessie/s3mock/S3Bucket.java [hunk 2]
```diff
@@ -47,11 +50,23 @@ public Lister lister() {
     return (String prefix) -> Stream.empty();
   }
 
+  @Value.Default
+  public PutObject putObject() {
```
Comment (adutra) [inline testing/s3mock/src/main/java/org/projectnessie/s3mock/S3Bucket.java:54]: Maybe rename to `storer` or `setter` to align with `lister` and `deleter`.

File: testing/s3mock/src/test/java/org/projectnessie/s3mock/TestIcebergS3MockServerS3Client.java [hunk 3]
```diff
@@ -318,4 +325,101 @@ public void getObject() {
...
+            .key("my-object")
+            .contentType("text/plain")
+            .build(),
+        RequestBody.fromBytes("Hello World".getBytes(StandardCharsets.UTF_8)));
+
+    soft.assertThat(
+            IoUtils.toByteArray(
+                s3.getObject(GetObjectRequest.builder().bucket(BUCKET).key("my-object").build())))
+        .asString()
+        .isEqualTo("Hello World");
```
Comment (adutra) [inline testing/s3mock/src/test/java/org/projectnessie/s3mock/TestIcebergS3MockServerS3Client.java:377]:
```suggestion
    soft.assertThat(
            IoUtils.toUtf8String(
                s3.getObject(GetObjectRequest.builder().bucket(BUCKET).key("my-object").build())))
        .isEqualTo("Hello World");
```

File: testing/s3mock/src/main/java/org/projectnessie/s3mock/AwsChunkedInputStream.java [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+          state = AwsChunkedState.EXPECT_HEADER;
+          break;
+        default:
+          throw new IllegalStateException();
+      }
+    }
+  }
+
+  @Override
+  public int read() throws IOException {
```
Comment (adutra) [inline testing/s3mock/src/main/java/org/projectnessie/s3mock/AwsChunkedInputStream.java:98]: This method doesn't appear to be covered by any test FYI.

File: testing/s3mock/src/main/java/org/projectnessie/s3mock/AwsChunkedInputStream.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+
+  // Interestingly, for a 1MB upload, the AWS sync client uses 'aws-chunked' content-encoding, but
+  // the AWS async client does not.
+
+  AwsChunkedInputStream(InputStream input) {
+    this.input = input;
+  }
+
+  enum AwsChunkedState {
+    EXPECT_HEADER,
```
Comment (adutra) [inline testing/s3mock/src/main/java/org/projectnessie/s3mock/AwsChunkedInputStream.java:38]: "Header" in this context is a bit misleading imo, maybe rename to `METADATA` to align with `DATA`.

File: testing/s3mock/src/main/java/org/projectnessie/s3mock/S3Bucket.java (hunk 2 above)
Comment (snazy) [reply testing/s3mock/src/main/java/org/projectnessie/s3mock/S3Bucket.java:54]: done

File: testing/s3mock/src/main/java/org/projectnessie/s3mock/AwsChunkedInputStream.java (hunk 4 above)
Comment (snazy) [reply testing/s3mock/src/main/java/org/projectnessie/s3mock/AwsChunkedInputStream.java:98]: added explicit test class

**Pull request merged**